gnat2goto/driver/tree_walk.adb

```diff
@@ -1472,15 +1472,28 @@ package body Tree_Walk is
    -----------------------------
 
    function Do_Qualified_Expression (N : Node_Id) return Irep is
-      Value : constant Irep := Do_Expression (Expression (N));
-      Type_Of_Val : constant Irep :=
-         Follow_Symbol_Type (
-          Do_Type_Reference (Etype (N)),
-          Global_Symbol_Table);
-      Typecast_Expr : constant Irep :=
-        Make_Op_Typecast (Value, Sloc (N), Type_Of_Val);
    begin
-      return Make_Range_Assert_Expr (N, Typecast_Expr, Type_Of_Val);
+      --  Todo qualified array expressions - at the moment they will not
+      --  be checked.
+      --  Only perform a range check if the qualification is for a numeric
+      --  enumeration or array subtype
+      if Is_Numeric_Type (Etype (N))
+        or else Is_Enumeration_Type (Etype (N))
+      then
+         declare
+            Value : constant Irep := Do_Expression (Expression (N));
+            Type_Of_Val : constant Irep :=
+              Follow_Symbol_Type (Do_Type_Reference (Etype (N)),
+                                  Global_Symbol_Table);
+            Typecast_Expr : constant Irep :=
+              Make_Op_Typecast (Value, Sloc (N), Type_Of_Val);
+         begin
+            return Make_Range_Assert_Expr (N, Typecast_Expr, Type_Of_Val);
+         end;
+      else
+         return Do_Expression (Expression (N));
```

Comment:
We might still want a type-cast here.

Reply:
It is not really a type cast but a check may be required for array and discriminated types to confirm that the expression is compatible with the given type qualifier.

I have added a "report unhandled node" when array or discriminated qualified expressions are encountered to warn that they are not checked.